cli/internal/codegen/testdata/expected_typescript.ts

```diff
@@ -385,34 +399,324 @@ class BaseClient {
         // Merge our headers with any predefined headers
         init.headers = {...this.headers, ...init.headers}
 
+        let bearerToken: string | undefined
+
         // If an authorization token generator is present, call it and add the returned token to the request
         if (this.tokenGenerator) {
-            init.headers["Authorization"] = "Bearer " + this.tokenGenerator()
+            bearerToken = this.tokenGenerator()
+        }
+
+        // If we now have a bearer token, add it to the request
+        if (bearerToken) {
+            init.headers["Authorization"] = "Bearer " + bearerToken
+        } else if (requiresAuth) {
+            // The API called requires authorization, but we don't have a token generator, or it provided a blank token value
+            throw new EncoreError(400, { code: ErrCode.Unauthenticated, message: 'Authorization token required for this API, but none provided.' })
         }
 
         // Make the actual request
         const response = await this.fetcher(this.baseURL + path, init)
 
         // handle any error responses
         if (!response.ok) {
-            const body = await response.text()
-            throw new Error(`request failed: status ${response.status}: ${body}`)
+            // try and get the error message from the response body
+            let body: EncoreErrorResponse = { code: ErrCode.Unknown, message: `request failed: status ${response.status}` }
+
+            // if we can get the structured error we should, otherwise give a best effort
+            try {
+                let jsonBody = await response.json()
+                if (isEncoreErrorResponse(jsonBody)) {
+                    body = jsonBody
+                } else {
+                    body.message += ": " + JSON.stringify(jsonBody)
+                }
+            } catch {
+                // otherwise we just append the text to the error message
+                body.message += ": " + await response.text()
+            }
+
+            throw new EncoreError(response.status, body)
         }
 
         return response
     }
 }
 
-function encodeQuery(parts: Record<string, any>): string {
-    const pairs = []
-    for (let key in parts) {
-        let val = parts[key]
-        if (!Array.isArray(val)) {
-            val = [val]
+/**
+ * EncoreErrorDetails represents the response from an Encore API in the case of an error
+ */
+interface EncoreErrorResponse {
+    code: ErrCode
+    message: string
+    details?: any
+}
+
+function isEncoreErrorResponse(err: any): err is EncoreErrorResponse {
+    return (
+        err !== undefined && err !== null && 
+        typeof(err.code) === "number" &&
+        typeof(err.message) === "string" &&
+        (err.details === undefined || err.details === null || typeof(err.details) === "object")
+    )
+}
+
+/**
+ * EncoreError represents a structured error as returned from an Encore application.
+ */
+export class EncoreError extends Error {
```

Comment:
Maybe this should be `APIError`? `EncoreError` might suggest it's an error **with** Encore or **from Encore the product/platform**, as opposed to from the *Encore application*.